gr-qtgui/examples/c++/display_qt.cc

```diff
@@ -0,0 +1,121 @@
+#include <gnuradio/fft/window.h>
+
+#include <QWidget>
+#include <QHBoxLayout>
+#include <QTabWidget>
+
+using namespace gr;
+
+int main(int argc, char **argv)
+{
+  // The global QT application
+  QApplication app(argc, argv);
+
+  // Main QT app widget
+  QWidget *main = new QWidget();
+
+  // We'll use a horizontal layout of two QTabWidgets
+  QHBoxLayout *layout = new QHBoxLayout();
+
+  // Create the tab widgets
+  QTabWidget *tab0 = new QTabWidget();
+  QTabWidget *tab1 = new QTabWidget();
+
+  // Add the tab widgets to the layou
+  layout->addWidget(tab0);
+  layout->addWidget(tab1);
+
+  // Set the layout as the main widget's layou
+  main->setLayout(layout);
+
+  // Simple resizing of the app
+  main->resize(2000,800);
+
+  // sample rate
+  int rate = 48000;
+
+  // Create the GNU Radio top block
+  top_block_sptr tb = make_top_block("display_qt");
+
+  // Source will be sine wave in noise
+  analog::sig_source_f::sptr src0 = analog::sig_source_f::make(rate, analog::GR_SIN_WAVE, 1500, 1);
+  analog::noise_source_f::sptr src1 = analog::noise_source_f::make(analog::GR_GAUSSIAN, 0.1);
+
+  // Combine signal and noise; add throttle
+  blocks::add_ff::sptr src = blocks::add_ff::make();
+  blocks::throttle::sptr thr = blocks::throttle::make(sizeof(float), rate);
+
+  // Create the QTGUI sinks
+  // Time, Freq, Waterfall, and Histogram sinks
+  qtgui::time_sink_f::sptr tsnk = qtgui::time_sink_f::make(1024, rate, "", 1);
+  qtgui::freq_sink_f::sptr fsnk = qtgui::freq_sink_f::make(1024, fft::window::WIN_HANN,
+                                                           0, rate, "", 1);
+  qtgui::waterfall_sink_f::sptr wsnk = qtgui::waterfall_sink_f::make(1024, fft::window::WIN_HANN,
+                                                                     0, rate, "", 1);
+  qtgui::histogram_sink_f::sptr hsnk = qtgui::histogram_sink_f::make(1024, 100, -2, 2, "", 1);
+
+  // Turn off the legend on these plots
+  tsnk->disable_legend();
+  fsnk->disable_legend();
+  hsnk->disable_legend();
+
+  // Connect the graph
+  tb->connect(src0, 0, src, 0);
+  tb->connect(src1, 0, src, 1);
+  tb->connect(src, 0, thr, 0);
+  tb->connect(thr, 0, tsnk, 0);
+  tb->connect(thr, 0, fsnk, 0);
+  tb->connect(thr, 0, wsnk, 0);
+  tb->connect(thr, 0, hsnk, 0);
+
+  // Get the raw QWidget objects from the GNU Radio blocks
+  QWidget* qtgui_time_sink_win = tsnk->qwidget();
+  QWidget* qtgui_freq_sink_win = fsnk->qwidget();
+  QWidget* qtgui_waterfall_sink_win = wsnk->qwidget();
+  QWidget* qtgui_histogram_sink_win = hsnk->qwidget();
+
+  // Plug the widgets into the tabs
+  tab0->addTab(qtgui_time_sink_win, "Time");
+  tab0->addTab(qtgui_histogram_sink_win, "Hist");
+  tab1->addTab(qtgui_freq_sink_win, "Freq");
+  tab1->addTab(qtgui_waterfall_sink_win, "Waterfall");
+
+  tb->start();   // Start the flowgraph
+  main->show();  // show the window
+  app.exec();    // run the QT executor loop
+  tb->stop();    // on closing the QT app, stop the flowgraph
```

Comment:
The stop() and wait() should really be done in a function called on the Qt "aboutToQuit" signal to make sure that the flow graph is stopped before Qt is done tearing itself down or some widget will try to redraw on free'd object and you can get a segfault on exit.